### PR TITLE
Issue/11 - Create integration test stages and cleanup workflows

### PR DESCRIPTION
**Pull request by nathandotleeathpe**

### CODE REVIEW

File: .github/workflows/pull_request.yml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+# PRs from forked repos will not have the permissions to invoke check runs
+# (i.e. publish test reports) or publish code coverage summary comments.
+# Instead, the "Publish Integration Test" and "Publish Unit Test" workflows
+# on the main branch will be activated using the "workflow_run" trigger. These
+# workflows will publish the test reports and code coverage summary comment on
+# behalf of the pull request run.
+#
+# NOTE: This workflow is designed to run on PRs from forked repositories and is
+# not guarenteed to run perfectly in PRs from branches in the same repository.
```
Comment (roehrich-hpe) [inline .github/workflows/pull_request.yml:30]: guarenteed => guaranteed

File: .github/workflows/publish-unit-test.yml [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+            let fs = require('fs');
+            fs.writeFileSync(`${process.env.GITHUB_WORKSPACE}/unit-test-results.zip`, Buffer.from(downloadTestResults.data));
+            fs.writeFileSync(`${process.env.GITHUB_WORKSPACE}/event.json.zip`, Buffer.from(downloadEventFile.data));
+
+      - name: Unzip artifacts
+        run: unzip unit-test-results.zip && unzip event.json.zip
+
+      # This action creates a checkrun that publishes the junit test report in
+      # the first workflow run in the commit. This should only be the 
+      # "Pull Request" workflow, but chould end up being the "Build" workflow
```
Comment (roehrich-hpe) [inline .github/workflows/publish-unit-test.yml:114]: chould => ?

File: .github/workflows/integration-test.yml [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+          push: false
+          load: true
+          tags: slurm-bb:test
+          build-args: |
+            SLURM_TAG=slurm-22-05-4-1
+          cache-from: type=gha,scope=slurm-${{ env.BRANCH }}
+          cache-to: type=gha,mode=max,scope=slurm-${{ env.BRANCH }}
+
+      # Pre-build slurm image with docker cache. Expect 3 minutes for a full
+      # DWS build.
```
Comment (roehrich-hpe) [inline .github/workflows/integration-test.yml:70]: When I have a release tag in DWS, you should just pull that build.

File: testsuite/integration/src/tests/test_environment.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    v1 = client.CoreV1Api()
+    return v1
+
+@pytest.fixture
+def slurmctld():
+    return docker.from_env().containers.get("slurmctld")
+
+@given('the kubernetes cluster kube-system UID', target_fixture="kube_system_uid")
+def _(k8s):
+    """the kubernetes cluster kube-system UID."""
```
[the remark this answers is not in the document]
Comment (nathandotleeathpe) [reply testsuite/integration/src/tests/test_environment.py:47]: pytest and pytest-bdd auto generate the decorators and function definitions, including the comments.

File: .github/workflows/integration-test.yml (hunk 3 above)
Comment (nathandotleeathpe) [reply .github/workflows/integration-test.yml:70]: Agreed.

File: testsuite/integration/Makefile [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+.PHONY: setup-kind
+setup-kind:
+	@{ \
+		set -e ; \
+		source kind/kind.sh ; \
+		test ! -z "$(SKIP_KIND_GENERATION)" && echo "Skipping cluster generation" || generate_cluster ; \
+		install_dependencies ; \
+		prep_kubeconfig ; \
+	}
+	cd slurm && docker compose build && docker compose up --wait
```
Comment (roehrich-hpe) [inline testsuite/integration/Makefile:34]: This "compose build" fails for me.  It looks like it's expecting to find a Dockerfile in the slurm/ dir, but you don't have one in there.

```
: dean-mbp; cd testsuite/integration/slurm 
: dean-mbp; ls
README.md		docker-compose.yml	jobs/
: dean-mbp; docker compose build
[+] Building 0.0s (1/2)                                                         
 => [internal] load build definition from Dockerfile                       0.0s
 => => transferring dockerfile: 2B                                         0.0s
failed to solve: rpc error: code = Unknown desc = failed to solve with frontend dockerfile.v0: failed to read dockerfile: open /var/lib/docker/tmp/buildkit-mount3097822521/Dockerfile: no such file or directory
```

Comment (roehrich-hpe) [reply testsuite/integration/Makefile:34]: > This "compose build" fails for me. It looks like it's expecting to find a Dockerfile in the slurm/ dir, but you don't have one in there.
> 
> ```
> : dean-mbp; cd testsuite/integration/slurm 
> : dean-mbp; ls
> README.md		docker-compose.yml	jobs/
> : dean-mbp; docker compose build
> [+] Building 0.0s (1/2)                                                         
>  => [internal] load build definition from Dockerfile                       0.0s
>  => => transferring dockerfile: 2B                                         0.0s
> failed to solve: rpc error: code = Unknown desc = failed to solve with frontend dockerfile.v0: failed to read dockerfile: open /var/lib/docker/tmp/buildkit-mount3097822521/Dockerfile: no such file or directory
> ```

Never mind.  This workarea was a fresh git-clone, and I had not done the submodule init.
It's fine.

File: testsuite/integration/Makefile [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+	docker network connect slurm_default dws-control-plane
+
+.PHONY: setup-dws
+setup-dws:
+	@{\
+		set -e ; \
+		cd ../submodules/dws ; \
+		docker buildx build -t local/dws-operator:test --cache-from="/tmp/.buildx-cache" --load . ; \
+		IMAGE_TAG_BASE=local/dws-operator VERSION=test KIND_CLUSTER=dws make kind-push deploy ; \
+		kubectl wait deployment -n dws-operator-system dws-operator-controller-manager --for condition=Available=True ; \
```
Comment (roehrich-hpe) [inline testsuite/integration/Makefile:44]: Add `--timeout=60s` to this wait.   In my experience, DWS often needs more than the default 30s.

**Comment by roehrich-hpe** [comment]
That's all I have to comment on so far.  I'm ready for more updates.

**Review by roehrich-hpe — APPROVED** [review]

**Pull request merged**